accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,185 @@
+
+```C#
+    public partial class Activity : IDisposable
+    {
+        public void AddTag(string key, int value) { }
+        public void AddTag(string key, double value) { }
+        public void AddTag(string key, bool value) { }
+        public void AddTag(string key, int [] value) { }
+        public void AddTag(string key, double [] value) { }
+        public void AddTag(string key, bool [] value) { } // how interesting this one?
+        public void AddTag(string key, string [] value) { }
+
+        // The following API is pending the approval in OT spec first before we add it in .NET.
+        public void AddTag(string key, ActivityTagsCollection value) { }
+
+        // We cannot change Tags property so we are introducing a new property with a new name
+        public ActivityTagsCollection TypedTags { get; }
+    }
+```
+
+***<u>OpenQuestion</u>***
+
+We need to decide about the behavior of the old property `Activity.Tags` when adding tags of non-string types values. The options are
+
+- Ignore all non-string typed values and return only the string type value tags.
+- Convert the non-string type values to string before returning it.
+
+I am inclining to the first option to avoid handling ToString with more complex types (e.g. arrays or ActivityTagsCollection).
+
+## Proposal Changing already exposed APIs in previous preview releases
+
+As we are introducing the ActivityTagsCollection, we can take advantage of that now and modify some of the exposed APIs to use this collection instead. Also, we are renaming Attributes to Tags for consistency inside BCL APIs.
+
+### ActivityEvent
+
+```C#
+    // Already exposed APIs
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public ActivityEvent(string name, DateTimeOffset timestamp, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>> Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly struct ActivityEvent
+    {
+        public ActivityEvent(string name, ActivityTagsCollection? tags)
+        public ActivityEvent(string name, DateTimeOffset timestamp, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
+    }
+```
+
+### ActivityLink
+
+```C#
+    // Already exposed APIs
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, IEnumerable<KeyValuePair<string, object>>? attributes)
+        public IEnumerable<KeyValuePair<string, object>>? Attributes { get; }
+    }
+
+    // change it to the following:
+    public readonly partial struct ActivityLink  : IEquatable<ActivityLink>
+    {
+        public ActivityLink(ActivityContext context, ActivityTagsCollection? tags)
+        public ActivityTagsCollection? Tags { get; }
```

Comment:
TypedTags instead of Tags

Reply:
ditto. Please let me know if you disagree and we can talk more about it.